redhat-developer/kam
# Executing kam command scenario

**What type of PR is this?**
/kind enhancement

**What does this PR do / why we need it**:
Scenario for creating kam bootstrap files
**Have you updated the necessary documentation?**


**Which issue(s) this PR fixes**:

Fixes #119, #127 

**How to test changes / Special notes to the reviewer**:
Test scenario should pass in CI
 
How to verify it locally:

1. login into the cluster as kubeadmin

2. Make sure that ```kam```, ```oc``` and ```gh``` binaries are in $PATH. 4.5+ cluster needs to be configured before launching the tests against it. The files ```kubeadmin-password``` and ```kubeconfig``` which contain cluster login details should be present in the ```auth``` directory and it should reside in the same directory as ```Makefile```. If it is not present in the ```auth``` directory, please create it. Then run ```make prepare-test-cluster``` to configure the ```4.5+``` cluster. ```make prepare-test-cluster``` comprises installation of sealed secrets, openshift pipelines, argocd operator and create sealed secrets instance.

3. Export these env variables
```
// https://github.com/redhat-developer/kam/blob/master/docs/journey/day1/prerequisites/service_repo.md#fork-the-taxi-app-repository
$ export SERVICE_REPO_URL=https://github.com/<username>/taxi

// Just provide a repo url in your GitHub user namespace
$ export GITOPS_REPO_URL=https://github.com/<username>/gitops.git

// https://github.com/redhat-developer/kam/blob/master/docs/journey/day1/prerequisites/quay.md
$ export IMAGE_REPO=quay.io/<username>/taxi
$ export DOCKERCONFIGJSON_PATH=/path/to/dockerconfig/file.json

// GitHub token
$ export GITHUB_TOKEN=<token>

// Paste the GitHub token in a file and export the file
$ export KAM_GITHUB_TOKEN_FILE=/path/to/github/token/file.txt

```

4. Run e2e test
```
$ make e2e
Log successfully started, logging into: /Users/amit/go/src/github.com/redhat-developer/kam/test/e2e/out/test-results/integration_2021-1-1_18-06-51.log
Running integration test in: /Users/amit/go/src/github.com/redhat-developer/kam/test/e2e/out/test-run
Working directory set to: /Users/amit/go/src/github.com/redhat-developer/kam/test/e2e/out/test-run
Before suite
gh command is :  [gh auth login --with-token]
The bash instance has been started and will be used for testing.
Before feature
Feature: Basic test
  Checks whether KAM top-level commands behave correctly.

  Scenario: KAM version                     # features/basic.feature:4
    When executing "kam version" succeeds   # shell.go:234 -> github.com/code-ready/clicumber/testsuite.ExecuteCommandSucceedsOrFails
    Then stderr should be empty             # shell.go:306 -> github.com/code-ready/clicumber/testsuite.CommandReturnShouldBeEmpty
    And stdout should contain "kam version" # shell.go:290 -> github.com/code-ready/clicumber/testsuite.CommandReturnShouldContain

  Scenario: Execute KAM bootstrap command without --push-to-git=true flag                                                                                                                                                                             # features/basic.feature:9
    When executing "kam bootstrap --service-repo-url $SERVICE_REPO_URL --gitops-repo-url $GITOPS_REPO_URL --image-repo $IMAGE_REPO --dockercfgjson $DOCKERCONFIGJSON_PATH --git-host-access-token $GITHUB_TOKEN --output bootstrapresources" succeeds # shell.go:234 -> github.com/code-ready/clicumber/testsuite.ExecuteCommandSucceedsOrFails
    Then stderr should be empty                                                                                                                                                                                                                       # shell.go:306 -> github.com/code-ready/clicumber/testsuite.CommandReturnShouldBeEmpty

  Scenario: KAM bootstrap command with --push-to-git=true and --overwrite flag                                                                                                                                                                                                       # features/basic.feature:13
    When executing "kam bootstrap --service-repo-url $SERVICE_REPO_URL --gitops-repo-url $GITOPS_REPO_URL --image-repo $IMAGE_REPO --dockercfgjson $DOCKERCONFIGJSON_PATH --git-host-access-token $GITHUB_TOKEN --output bootstrapresources --push-to-git=true --overwrite" succeeds # shell.go:234 -> github.com/code-ready/clicumber/testsuite.ExecuteCommandSucceedsOrFails
    Then stderr should be empty                                                                                                                                                                                                                                                      # shell.go:306 -> github.com/code-ready/clicumber/testsuite.CommandReturnShouldBeEmpty
After feature
After suite
gh command is :  [gh alias set repo-delete api -X DELETE "repos/$1"]
gh command is :  [gh repo-delete amitkrout/gitops]

3 scenarios (3 passed)
7 steps (7 passed)
18.537791515s
ok      github.com/redhat-developer/kam/test/e2e        18.984s
```
